File: src/main/java/com/wantedmarket/member/service/MemberService.java
package com.wantedmarket.member.service;

import com.wantedmarket.member.domain.Member;
import com.wantedmarket.member.exception.MemberException;
import com.wantedmarket.member.repository.MemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import static com.wantedmarket.member.exception.MemberErrorCode.NO_MEMBER;

@Service
@RequiredArgsConstructor
public class MemberService {

    private final MemberRepository memberRepository;

    public Member getMember(String username) {
        return memberRepository.findByUsername(username)
                .orElseThrow(() -> new MemberException(NO_MEMBER));
    }

    public void save(Member member) {
        memberRepository.save(member);
    }

    public boolean existMember(String username) {
        return memberRepository.existsByUsername(username);
    }
}

File: src/main/kotlin/wanted/market/member/service/MemberService.kt
package wanted.market.member.service

import jakarta.transaction.Transactional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import wanted.market.common.exception.ErrorCode
import wanted.market.member.exception.MemberException
import wanted.market.member.dto.SaveMemberRequest
import wanted.market.member.entity.Member
import wanted.market.member.repository.MemberRepository

@Service
@Transactional
class MemberService(
    @Autowired private val memberRepository: MemberRepository
) {
    fun save(saveMemberRequest: SaveMemberRequest) {
        checkDuplication(saveMemberRequest.email)
        joinMember(saveMemberRequest)
    }

    private fun joinMember(saveMemberRequest: SaveMemberRequest) {
        val member = Member(
            name = saveMemberRequest.name,
            email = saveMemberRequest.email
        )
        memberRepository.save(member)
    }

    private fun checkDuplication(email: String) {
        val existMember = memberRepository.findByEmail(email)
        if (existMember.isPresent) {
            throw MemberException(ErrorCode.DUPLICATE_EMAIL)
        }
    }

    fun findMember(memberId: Long): Member {
        val member = memberRepository.findById(memberId)
            .orElseThrow{ MemberException(ErrorCode.MEMBER_NOT_FOUND) }
        return member
    }
}